include/libtorrent/aux_/rtc_stream.hpp

```diff
@@ -0,0 +1,391 @@
+	~rtc_stream_impl() = default;
+	rtc_stream_impl& operator=(rtc_stream_impl const&) = delete;
+	rtc_stream_impl(rtc_stream_impl const&) = delete;
+	rtc_stream_impl& operator=(rtc_stream_impl&&) noexcept = delete;
+	rtc_stream_impl(rtc_stream_impl&&) noexcept = delete;
+
+	void init();
+	void close();
+
+	bool is_open() const;
+	std::size_t available() const;
+
+	endpoint_type local_endpoint(error_code& ec) const;
+	endpoint_type remote_endpoint(error_code& ec) const;
+
+	void cancel_handlers(error_code const& ec);
+
+	bool has_read_handler() const { return bool(m_read_handler); }
+	bool has_write_handler() const { return bool(m_write_handler); }
+	template <class Handler>
+	void set_read_handler(Handler handler) { m_read_handler = std::move(handler); }
+	template <class Handler>
+	void set_write_handler(Handler handler) { m_write_handler = std::move(handler); }
+
+	template <class Mutable_Buffer>
+	std::size_t add_read_buffer(Mutable_Buffer const& buffer) {
+		if(buffer.size() == 0) return 0;
+		m_read_buffer.push_back(buffer);
+		m_read_buffer_size += buffer.size();
+		return buffer.size();
+	}
+
+	template <class Const_Buffer>
+	std::size_t add_write_buffer(Const_Buffer const& buffer) {
+		if(buffer.size() == 0) return 0;
+		m_write_buffer.push_back(buffer);
+		m_write_buffer_size += buffer.size();
+		return buffer.size();
+	}
+
+	void issue_read();
+	void issue_write();
+	std::size_t read_some(error_code& ec);
+	std::size_t write_some(error_code& ec);
+	void clear_read_buffers();
+	void clear_write_buffers();
+
+private:
+	void on_available(error_code const& ec);
+	void on_buffered_low(error_code const& ec);
+	bool ensure_open();
+
+	static endpoint_type parse_endpoint(std::string const& addr, error_code& ec);
+
+	std::size_t incoming_data(span<char const> data);
+	std::pair<std::size_t, bool> write_data(std::size_t size);
+
+	io_context& m_io_context;
+	std::shared_ptr<rtc::PeerConnection> m_peer_connection;
+	std::shared_ptr<rtc::DataChannel> m_data_channel;
+
+	std::function<void(error_code const&, std::size_t)> m_read_handler;
+	std::function<void(error_code const&, std::size_t)> m_write_handler;
+	std::list<boost::asio::const_buffer> m_write_buffer;
+	std::list<boost::asio::mutable_buffer> m_read_buffer;
+	std::size_t m_write_buffer_size = 0;
+	std::size_t m_read_buffer_size = 0;
+
+	std::vector<char> m_incoming;
+};
+
+// This is the user-level stream interface to WebRTC DataChannels.
+struct TORRENT_EXTRA_EXPORT rtc_stream
+{
+	using lowest_layer_type = rtc_stream;
+	using endpoint_type = rtc_stream_impl::endpoint_type;
+	using protocol_type = rtc_stream_impl::protocol_type;
+
+	using executor_type = tcp::socket::executor_type;
+	executor_type get_executor() { return m_io_context.get_executor(); }
+
+	rtc_stream(io_context& ioc, rtc_stream_init init);
+	~rtc_stream();
+	rtc_stream& operator=(rtc_stream const&) = delete;
+	rtc_stream(rtc_stream const&) = delete;
+	rtc_stream& operator=(rtc_stream&&) noexcept = delete;
```

Comment:
I noticed you changed this from `= default` to `= delete`. was that deliberate? or did you mean to just remove `noexcept`? There are a fair number of types in STL that are nothrow-move-constructible, but still not nothrow-move-assignable.

Reply:
I was deliberate to make the implementation object non movable, as it should only be manipulated via pointer by `rtc_stream`.

Reply:
it's still move-constructible though (see the line below)